diff -ru binutils-orig/bfd/archures.c binutils-2.19.1/bfd/archures.c
--- binutils-orig/bfd/archures.c	2023-11-09 10:22:30.151989077 +0000
+++ binutils-2.19.1/bfd/archures.c	2023-11-09 13:41:40.791744425 +0000
@@ -407,6 +407,8 @@
 .#define bfd_mach_z80            3 {* With ixl, ixh, iyl, and iyh.  *}
 .#define bfd_mach_z80full        7 {* All undocumented instructions.  *}
 .#define bfd_mach_r800           11 {* R800: successor with multiplication.  *}
+.  bfd_arch_tms9900,   {* Texas Instruments TMS9900 archetecture.  *}
+.#define bfd_mach_tms9900       1
 .  bfd_arch_last
 .  };
 */
@@ -501,6 +503,7 @@
 extern const bfd_arch_info_type bfd_tic4x_arch;
 extern const bfd_arch_info_type bfd_tic54x_arch;
 extern const bfd_arch_info_type bfd_tic80_arch;
+extern const bfd_arch_info_type bfd_tms9900_arch;
 extern const bfd_arch_info_type bfd_v850_arch;
 extern const bfd_arch_info_type bfd_vax_arch;
 extern const bfd_arch_info_type bfd_we32k_arch;
@@ -570,6 +573,7 @@
     &bfd_tic4x_arch,
     &bfd_tic54x_arch,
     &bfd_tic80_arch,
+    &bfd_tms9900_arch,
     &bfd_v850_arch,
     &bfd_vax_arch,
     &bfd_w65_arch,
diff -ru binutils-orig/bfd/bfd-in2.h binutils-2.19.1/bfd/bfd-in2.h
--- binutils-orig/bfd/bfd-in2.h	2023-11-09 10:22:30.155989145 +0000
+++ binutils-2.19.1/bfd/bfd-in2.h	2023-11-09 13:41:40.795744410 +0000
@@ -2029,6 +2029,8 @@
 #define bfd_mach_z80            3 /* With ixl, ixh, iyl, and iyh.  */
 #define bfd_mach_z80full        7 /* All undocumented instructions.  */
 #define bfd_mach_r800           11 /* R800: successor with multiplication.  */
+  bfd_arch_tms9900,   /* Texas Instruments TMS9900 archetecture.  */
+#define bfd_mach_tms9900       1
   bfd_arch_last
   };
 
diff -ru binutils-orig/bfd/config.bfd binutils-2.19.1/bfd/config.bfd
--- binutils-orig/bfd/config.bfd	2023-11-09 10:22:30.159989213 +0000
+++ binutils-2.19.1/bfd/config.bfd	2023-11-09 13:41:40.795744410 +0000
@@ -102,6 +102,7 @@
 spu*)            targ_archs=bfd_spu_arch ;;
 strongarm*)	 targ_archs=bfd_arm_arch ;;
 thumb*)		 targ_archs=bfd_arm_arch ;;
+tms9900*)	 targ_archs=bfd_tms9900_arch ;;
 v850*)		 targ_archs=bfd_v850_arch ;;
 x86_64*)	 targ_archs=bfd_i386_arch ;;
 xscale*)	 targ_archs=bfd_arm_arch ;;
@@ -1432,6 +1433,10 @@
     targ_underscore=yes
     ;;
 
+  tms9900-*-*)
+    targ_defvec=bfd_elf32_tms9900_vec
+    ;;
+
   v850-*-*)
     targ_defvec=bfd_elf32_v850_vec
     ;;
diff -ru binutils-orig/bfd/configure binutils-2.19.1/bfd/configure
--- binutils-orig/bfd/configure	2023-11-09 10:22:30.167989348 +0000
+++ binutils-2.19.1/bfd/configure	2023-11-09 13:41:40.807744365 +0000
@@ -19740,6 +19740,7 @@
     bfd_elf32_sparc_vec)	tb="$tb elf32-sparc.lo elfxx-sparc.lo elf-vxworks.lo elf32.lo $elf" ;;
     bfd_elf32_sparc_vxworks_vec) tb="$tb elf32-sparc.lo elfxx-sparc.lo elf-vxworks.lo elf32.lo $elf" ;;
     bfd_elf32_spu_vec)		tb="$tb elf32-spu.lo elf32.lo $elf" ;;
+    bfd_elf32_tms9900_vec)	tb="$tb elf32-tms9900.lo elf32.lo $elf" ;;
     bfd_elf32_tradbigmips_vec)  tb="$tb elf32-mips.lo elfxx-mips.lo elf-vxworks.lo elf32.lo $elf ecofflink.lo" ;;
     bfd_elf32_tradlittlemips_vec) tb="$tb elf32-mips.lo elfxx-mips.lo elf-vxworks.lo elf32.lo $elf ecofflink.lo" ;;
     bfd_elf32_us_cris_vec)	tb="$tb elf32-cris.lo elf32.lo $elf" ;;
diff -ru binutils-orig/bfd/doc/archures.texi binutils-2.19.1/bfd/doc/archures.texi
--- binutils-orig/bfd/doc/archures.texi	2023-11-09 10:22:30.171989416 +0000
+++ binutils-2.19.1/bfd/doc/archures.texi	2023-11-09 13:41:40.807744365 +0000
@@ -372,6 +372,8 @@
 #define bfd_mach_z80            3 /* With ixl, ixh, iyl, and iyh.  */
 #define bfd_mach_z80full        7 /* All undocumented instructions.  */
 #define bfd_mach_r800           11 /* R800: successor with multiplication.  */
+  bfd_arch_tms9900,   /* Texas Instruments TMS9900 archetecture.  */
+#define bfd_mach_tms9900       1
   bfd_arch_last
   @};
 @end example
diff -ru binutils-orig/bfd/Makefile.in binutils-2.19.1/bfd/Makefile.in
--- binutils-orig/bfd/Makefile.in	2023-11-09 10:22:30.151989077 +0000
+++ binutils-2.19.1/bfd/Makefile.in	2023-11-09 13:41:40.791744425 +0000
@@ -546,6 +546,7 @@
 	elfxx-sparc.lo \
 	elf32-sparc.lo \
 	elf32-spu.lo \
+	elf32-tms9900.lo\
 	elf32-v850.lo \
 	elf32-vax.lo \
 	elf32-xstormy16.lo \
@@ -2140,6 +2141,9 @@
   $(INCDIR)/bfdlink.h $(INCDIR)/hashtab.h elf-bfd.h $(INCDIR)/elf/common.h \
   $(INCDIR)/elf/external.h $(INCDIR)/elf/internal.h $(INCDIR)/elf/spu.h \
   $(INCDIR)/elf/reloc-macros.h elf32-spu.h elf32-target.h
+elf32-tms9900.lo: elf32-tms9900.c $(INCDIR)/filenames.h $(INCDIR)/bfdlink.h\
+  $(INCDIR)/elf/external.h $(INCDIR)/elf/internal.h $(INCDIR)/elf/tms9900.h \
+  $(INCDIR)/elf/reloc-macros.h elf32-spu.h elf32-target.h
 elf32-v850.lo: elf32-v850.c $(INCDIR)/filenames.h $(INCDIR)/bfdlink.h \
   $(INCDIR)/hashtab.h elf-bfd.h $(INCDIR)/elf/common.h \
   $(INCDIR)/elf/external.h $(INCDIR)/elf/internal.h $(INCDIR)/elf/v850.h \
diff -ru binutils-orig/bfd/targets.c binutils-2.19.1/bfd/targets.c
--- binutils-orig/bfd/targets.c	2023-11-09 10:22:30.171989416 +0000
+++ binutils-2.19.1/bfd/targets.c	2023-11-09 13:41:40.811744350 +0000
@@ -660,6 +660,7 @@
 extern const bfd_target bfd_elf32_sparc_vec;
 extern const bfd_target bfd_elf32_sparc_vxworks_vec;
 extern const bfd_target bfd_elf32_spu_vec;
+extern const bfd_target bfd_elf32_tms9900_vec;
 extern const bfd_target bfd_elf32_tradbigmips_vec;
 extern const bfd_target bfd_elf32_tradlittlemips_vec;
 extern const bfd_target bfd_elf32_us_cris_vec;
diff -ru binutils-orig/binutils/dwarf.c binutils-2.19.1/binutils/dwarf.c
--- binutils-orig/binutils/dwarf.c	2023-11-09 10:22:30.171989416 +0000
+++ binutils-2.19.1/binutils/dwarf.c	2023-11-09 13:41:40.811744350 +0000
@@ -53,7 +53,7 @@
 int do_debug_macinfo;
 int do_debug_str;
 int do_debug_loc;
-int do_wide;
+static int do_wide;
 
 dwarf_vma (*byte_get) (unsigned char *, int);
 
diff -ru binutils-orig/binutils/readelf.c binutils-2.19.1/binutils/readelf.c
--- binutils-orig/binutils/readelf.c	2023-11-09 10:22:30.175989484 +0000
+++ binutils-2.19.1/binutils/readelf.c	2023-11-09 13:41:40.815744334 +0000
@@ -152,6 +152,7 @@
 #include "elf/sh.h"
 #include "elf/sparc.h"
 #include "elf/spu.h"
+#include "elf/tms9900.h"
 #include "elf/v850.h"
 #include "elf/vax.h"
 #include "elf/x86-64.h"
@@ -165,7 +166,7 @@
 #include "safe-ctype.h"
 
 char *program_name = "readelf";
-int do_wide;
+static int do_wide;
 static long archive_file_offset;
 static unsigned long archive_file_size;
 static unsigned long dynamic_addr;
@@ -1198,6 +1199,10 @@
 	case EM_CR16_OLD:
 	  rtype = elf_cr16_reloc_type (type);
 	  break;
+
+        case EM_TMS9900:
+          rtype = elf_tms9900_reloc_type (type);
+          break;
 	}
 
       if (rtype == NULL)
@@ -1825,6 +1830,7 @@
     case EM_CYGNUS_MEP:         return "Toshiba MeP Media Engine";
     case EM_CR16:		
     case EM_CR16_OLD:		return "National Semiconductor's CR16";
+    case EM_TMS9900:            return "Texas Instruments TMS9900";
     default:
       snprintf (buff, sizeof (buff), _("<unknown>: 0x%x"), e_machine);
       return buff;
diff -ru binutils-orig/config.sub binutils-2.19.1/config.sub
--- binutils-orig/config.sub	2023-11-09 10:22:30.175989484 +0000
+++ binutils-2.19.1/config.sub	2023-11-09 13:41:40.815744334 +0000
@@ -148,7 +148,7 @@
 	-convergent* | -ncr* | -news | -32* | -3600* | -3100* | -hitachi* |\
 	-c[123]* | -convex* | -sun | -crds | -omron* | -dg | -ultra | -tti* | \
 	-harris | -dolphin | -highlevel | -gould | -cbm | -ns | -masscomp | \
-	-apple | -axis | -knuth | -cray)
+	-apple | -axis | -knuth | -cray | -tms9900)
 		os=
 		basic_machine=$1
 		;;
@@ -284,7 +284,7 @@
 	| sparc | sparc64 | sparc64b | sparc64v | sparc86x | sparclet | sparclite \
 	| sparcv8 | sparcv9 | sparcv9b | sparcv9v \
 	| spu | strongarm \
-	| tahoe | thumb | tic4x | tic80 | tron \
+	| tahoe | thumb | tic4x | tic80 | tron | tms9900 \
 	| v850 | v850e \
 	| we32k \
 	| x86 | xc16x | xscale | xscalee[bl] | xstormy16 | xtensa \
diff -ru binutils-orig/gas/configure.tgt binutils-2.19.1/gas/configure.tgt
--- binutils-orig/gas/configure.tgt	2023-11-09 10:22:30.175989484 +0000
+++ binutils-2.19.1/gas/configure.tgt	2023-11-09 13:41:40.815744334 +0000
@@ -81,6 +81,7 @@
   strongarm*be)		cpu_type=arm endian=big ;;
   strongarm*b)		cpu_type=arm endian=big ;;
   strongarm*)		cpu_type=arm endian=little ;;
+  tsm9900*)             cpu_type=tms9900 endian=big ;;
   v850*)		cpu_type=v850 ;;
   x86_64*)		cpu_type=i386 arch=x86_64;;
   xscale*be|xscale*b)	cpu_type=arm endian=big ;;
@@ -384,6 +385,8 @@
   tic4x-*-* | c4x-*-*)			fmt=coff bfd_gas=yes ;;
   tic54x-*-* | c54x*-*-*)		fmt=coff bfd_gas=yes need_libm=yes;;
 
+  tms9900-*-*)				fmt=elf ;;
+
   v850-*-*)				fmt=elf ;;
   v850e-*-*)				fmt=elf ;;
   v850ea-*-*)				fmt=elf ;;
diff -ru binutils-orig/gas/dwarf2dbg.h binutils-2.19.1/gas/dwarf2dbg.h
--- binutils-orig/gas/dwarf2dbg.h	2023-11-09 10:22:30.175989484 +0000
+++ binutils-2.19.1/gas/dwarf2dbg.h	2023-11-09 13:41:40.815744334 +0000
@@ -81,12 +81,12 @@
 
 /* True when we've seen a .loc directive recently.  Used to avoid
    doing work when there's nothing to do.  */
-bfd_boolean dwarf2_loc_directive_seen;
+extern bfd_boolean dwarf2_loc_directive_seen;
 
 /* True when we're supposed to set the basic block mark whenever a label
    is seen.  Unless the target is doing Something Weird, just call 
    dwarf2_emit_label.  */
-bfd_boolean dwarf2_loc_mark_labels;
+extern bfd_boolean dwarf2_loc_mark_labels;
 
 extern void dwarf2_finish (void);
 
diff -ru binutils-orig/gprof/hist.c binutils-2.19.1/gprof/hist.c
--- binutils-orig/gprof/hist.c	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/gprof/hist.c	2023-11-09 13:41:40.815744334 +0000
@@ -77,6 +77,9 @@
   { 'a', 1e+18 }				/* ato */
 };
 
+histogram *histograms;
+unsigned num_histograms;
+
 /* Reads just the header part of histogram record into
    *RECORD from IFP.  FILENAME is the name of IFP and
    is provided for formatting error messages only.  
diff -ru binutils-orig/gprof/hist.h binutils-2.19.1/gprof/hist.h
--- binutils-orig/gprof/hist.h	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/gprof/hist.h	2023-11-09 13:41:40.815744334 +0000
@@ -30,8 +30,8 @@
   int *sample;           /* Histogram samples (shorts in the file!).  */
 } histogram;
 
-histogram *histograms;
-unsigned num_histograms;
+extern histogram *histograms;
+extern unsigned num_histograms;
 
 /* Scale factor converting samples to pc values:
    each sample covers HIST_SCALE bytes.  */
diff -ru binutils-orig/include/dis-asm.h binutils-2.19.1/include/dis-asm.h
--- binutils-orig/include/dis-asm.h	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/include/dis-asm.h	2023-11-09 13:41:40.815744334 +0000
@@ -275,6 +275,7 @@
 extern int print_insn_tic4x		(bfd_vma, disassemble_info *);
 extern int print_insn_tic54x		(bfd_vma, disassemble_info *);
 extern int print_insn_tic80		(bfd_vma, disassemble_info *);
+extern int print_insn_tms9900           (bfd_vma, disassemble_info *);
 extern int print_insn_v850		(bfd_vma, disassemble_info *);
 extern int print_insn_vax		(bfd_vma, disassemble_info *);
 extern int print_insn_w65		(bfd_vma, disassemble_info *);
diff -ru binutils-orig/include/elf/common.h binutils-2.19.1/include/elf/common.h
--- binutils-orig/include/elf/common.h	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/include/elf/common.h	2023-11-09 13:41:40.815744334 +0000
@@ -334,6 +334,7 @@
 
 #define EM_CYGNUS_MEP		0xF00D  /* Toshiba MeP */
 
+#define EM_TMS9900		0xABBA
 /* See the above comment before you add a new EM_* value here.  */
 
 /* Values for e_version.  */
diff -ru binutils-orig/ld/configure.tgt binutils-2.19.1/ld/configure.tgt
--- binutils-orig/ld/configure.tgt	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/ld/configure.tgt	2023-11-09 13:41:40.819744320 +0000
@@ -607,6 +607,8 @@
 tic54x-*-* | c54x*-*-*)	targ_emul=tic54xcoff ;;
 tic80-*-*)		targ_emul=tic80coff
 			;;
+tms9900-*-*)		targ_emul=tms9900
+			;;
 v850-*-*)		targ_emul=v850 ;;
 v850e-*-*)		targ_emul=v850 ;;
 v850ea-*-*)		targ_emul=v850
diff -ru binutils-orig/ld/Makefile.in binutils-2.19.1/ld/Makefile.in
--- binutils-orig/ld/Makefile.in	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/ld/Makefile.in	2023-11-09 13:41:40.819744320 +0000
@@ -640,6 +640,7 @@
 	etic4xcoff.o \
 	etic54xcoff.o \
 	etic80coff.o \
+	etms9900.o \
 	evanilla.o \
 	evax.o \
 	evaxnbsd.o \
@@ -2520,6 +2521,9 @@
 etic80coff.c: $(srcdir)/emulparams/tic80coff.sh \
   $(srcdir)/emultempl/generic.em $(srcdir)/scripttempl/tic80coff.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} tic80coff "$(tdir_tic80coff)"
+etms9900.c: $(srcdir)/emulparams/tms9900.sh \
+  $(ELF_DEPS) $(srcdir)/emultempl/generic.em $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} tms9900 "$(tdir_tms9900)"
 evanilla.c: $(srcdir)/emulparams/vanilla.sh \
   $(srcdir)/emultempl/vanilla.em $(srcdir)/scripttempl/vanilla.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} vanilla "$(tdir_vanilla)"
diff -ru binutils-orig/opcodes/configure binutils-2.19.1/opcodes/configure
--- binutils-orig/opcodes/configure	2023-11-09 10:22:30.183989619 +0000
+++ binutils-2.19.1/opcodes/configure	2023-11-09 13:41:40.819744320 +0000
@@ -11885,6 +11885,7 @@
         bfd_tic4x_arch)         ta="$ta tic4x-dis.lo" ;;
 	bfd_tic54x_arch)	ta="$ta tic54x-dis.lo tic54x-opc.lo" ;;
 	bfd_tic80_arch)		ta="$ta tic80-dis.lo tic80-opc.lo" ;;
+        bfd_tms9900_arch)	ta="$ta tms9900-dis.lo" ;;
 	bfd_v850_arch)		ta="$ta v850-opc.lo v850-dis.lo" ;;
 	bfd_v850e_arch)		ta="$ta v850-opc.lo v850-dis.lo" ;;
 	bfd_v850ea_arch)	ta="$ta v850-opc.lo v850-dis.lo" ;;
diff -ru binutils-orig/opcodes/disassemble.c binutils-2.19.1/opcodes/disassemble.c
--- binutils-orig/opcodes/disassemble.c	2023-11-09 10:22:30.183989619 +0000
+++ binutils-2.19.1/opcodes/disassemble.c	2023-11-09 13:41:40.819744320 +0000
@@ -77,6 +77,7 @@
 #define ARCH_tic4x
 #define ARCH_tic54x
 #define ARCH_tic80
+#define ARCH_tms9900
 #define ARCH_v850
 #define ARCH_vax
 #define ARCH_w65
@@ -386,6 +387,11 @@
       disassemble = print_insn_tic80;
       break;
 #endif
+#ifdef ARCH_tms9900
+    case bfd_arch_tms9900:
+      disassemble = print_insn_tms9900;
+      break;
+#endif
 #ifdef ARCH_v850
     case bfd_arch_v850:
       disassemble = print_insn_v850;
diff -ru binutils-orig/opcodes/Makefile.in binutils-2.19.1/opcodes/Makefile.in
--- binutils-orig/opcodes/Makefile.in	2023-11-09 10:22:30.179989552 +0000
+++ binutils-2.19.1/opcodes/Makefile.in	2023-11-09 13:41:40.819744320 +0000
@@ -420,7 +420,8 @@
 	tic54x-opc.c \
 	tic80-dis.c \
 	tic80-opc.c \
-	v850-dis.c \
+	tms9900-dis.c \
+        v850-dis.c \
 	v850-opc.c \
 	vax-dis.c \
 	w65-dis.c \
@@ -1664,6 +1665,8 @@
   $(INCDIR)/symcat.h
 tic80-opc.lo: tic80-opc.c sysdep.h config.h $(INCDIR)/ansidecl.h \
   $(INCDIR)/opcode/tic80.h
+tms9900-dis.lo: tms9900-dis.c sysdep.h config.h $(INCDIR)/ansidecl.h \
+  $(INCDIR)/dis-asm.h
 v850-dis.lo: v850-dis.c sysdep.h config.h $(INCDIR)/ansidecl.h \
   $(INCDIR)/opcode/v850.h $(INCDIR)/dis-asm.h $(BFD_H) \
   $(INCDIR)/symcat.h opintl.h
