diff -rupN binutils-2.19.1/binutils/dwarf.c binutils-2.19.1-tms9900/binutils/dwarf.c
--- binutils-2.19.1/binutils/dwarf.c	2021-01-07 08:17:12.000000000 -0600
+++ binutils-2.19.1-tms9900/binutils/dwarf.c	2021-01-07 08:19:18.000000000 -0600
@@ -53,7 +53,7 @@
 int do_debug_macinfo;
 int do_debug_str;
 int do_debug_loc;
-int do_wide;
+static int do_wide;
 
 dwarf_vma (*byte_get) (unsigned char *, int);
 
diff -rupN binutils-2.19.1/binutils/readelf.c binutils-2.19.1-tms9900/binutils/readelf.c
--- binutils-2.19.1/binutils/readelf.c	2021-01-07 08:17:23.000000000 -0600
+++ binutils-2.19.1-tms9900/binutils/readelf.c	2021-01-07 08:19:10.000000000 -0600
@@ -166,7 +166,7 @@
 #include "safe-ctype.h"
 
 char *program_name = "readelf";
-int do_wide;
+static int do_wide;
 static long archive_file_offset;
 static unsigned long archive_file_size;
 static unsigned long dynamic_addr;
diff -rupN binutils-2.19.1/gas/dwarf2dbg.h binutils-2.19.1-tms9900/gas/dwarf2dbg.h
--- binutils-2.19.1/gas/dwarf2dbg.h	2021-01-07 08:33:33.000000000 -0600
+++ binutils-2.19.1-tms9900/gas/dwarf2dbg.h	2021-01-07 08:40:07.000000000 -0600
@@ -81,12 +81,12 @@
 
 /* True when we've seen a .loc directive recently.  Used to avoid
    doing work when there's nothing to do.  */
-bfd_boolean dwarf2_loc_directive_seen;
+extern bfd_boolean dwarf2_loc_directive_seen;
 
 /* True when we're supposed to set the basic block mark whenever a label
    is seen.  Unless the target is doing Something Weird, just call 
    dwarf2_emit_label.  */
-bfd_boolean dwarf2_loc_mark_labels;
+extern bfd_boolean dwarf2_loc_mark_labels;
 
 extern void dwarf2_finish (void);
 
diff -rupN binutils-2.19.1/gprof/hist.c binutils-2.19.1-tms9900/gprof/hist.c
--- binutils-2.19.1/gprof/hist.c	2021-01-07 08:48:31.000000000 -0600
+++ binutils-2.19.1-tms9900/gprof/hist.c	2021-01-07 08:51:29.000000000 -0600
@@ -77,6 +77,9 @@
   { 'a', 1e+18 }				/* ato */
 };
 
+histogram *histograms;
+unsigned num_histograms;
+
 /* Reads just the header part of histogram record into
    *RECORD from IFP.  FILENAME is the name of IFP and
    is provided for formatting error messages only.  
diff -rupN binutils-2.19.1/gprof/hist.h binutils-2.19.1-tms9900/gprof/hist.h
--- binutils-2.19.1/gprof/hist.h	2021-01-07 08:47:56.000000000 -0600
+++ binutils-2.19.1-tms9900/gprof/hist.h	2021-01-07 08:48:25.000000000 -0600
@@ -29,8 +29,8 @@
   int *sample;           /* Histogram samples (shorts in the file!).  */
 } histogram;
 
-histogram *histograms;
-unsigned num_histograms;
+extern histogram *histograms;
+extern unsigned num_histograms;
 
 /* Scale factor converting samples to pc values:
    each sample covers HIST_SCALE bytes.  */
